### PR TITLE
Add test for HJO&HBO when spill fails

### PR DESCRIPTION
**Pull request by szymonm**

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 1]
```diff
@@ -96,7 +96,7 @@
 {
     private static final int PARTITION_COUNT = 4;
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java:104]: cmt "Add partitioningSpillerFactory parameter to innerJoinOperatorFactory" -- consider mentioning "test" in cmt msg, e.g. "Add partitioningSpillerFactory parameter to test innerJoinOperatorFactory"

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 2]
```diff
@@ -96,7 +96,7 @@
...
     private static final LookupJoinOperators LOOKUP_JOIN_OPERATORS = new LookupJoinOperators(new JoinProbeCompiler());
-    private static final SingleStreamSpillerFactory SINGLE_STREAM_SPILLER_FACTORY = new DummySpillerFactory();
+    private static final DummySpillerFactory SINGLE_STREAM_SPILLER_FACTORY = new DummySpillerFactory();
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: looks like change unrelated to "Add partitioningSpillerFactory parameter to innerJoinOperatorFactory" cmt

File: presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java [hunk 3]
```diff
@@ -78,6 +99,14 @@ private Builder(Executor notificationExecutor, ScheduledExecutorService yieldExe
             this.session = session;
         }
 
+        public Builder(Executor notificationExecutor, ScheduledExecutorService yieldExecutor, Session session, TaskStateMachine taskStateMachine)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java:102]: Builder that takes all (or almost all) required parameters immediately in the ctor, is kind of not a builder, is it?

File: presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java [hunk 4]
```diff
@@ -123,7 +152,12 @@ public TaskContext build()
                     queryMaxSpillSize,
                     spillSpaceTracker);
 
-            return createTaskContext(queryContext, notificationExecutor, session);
+            if (taskStateMachine == null) {
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java:155]: This class already knows how to construct `TaskStateMachine` (`new TaskStateMachine(new TaskId("query", 0, 0), executor)`), so it could use this as default and let callers override. This way you could avoid the `if` in `build()`.

File: presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java [hunk 5]
```diff
@@ -46,6 +46,12 @@ public static TaskContext createTaskContext(Executor notificationExecutor, Sched
                 .build();
     }
 
+    public static TaskContext createTaskContext(Executor notificationExecutor, ScheduledExecutorService yieldExecutor, Session session, TaskStateMachine taskStateMachine)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java:49]: Maybe instead of adding new overload, make it possible to achieve same result using `builder()`?

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 6]
```diff
@@ -231,17 +238,63 @@ public void testYield()
 
     @Test(dataProvider = "testInnerJoinWithSpillDataProvider")
     public void testInnerJoinWithSpill(boolean probeHashEnabled, List<WhenSpill> whenSpill)
-            throws Exception
+            throws Throwable
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: You can revert to `throws Exception` if you apply one of further comments.

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 7]
```diff
@@ -283,18 +337,38 @@ public void testInnerJoinWithSpill(boolean probeHashEnabled, List<WhenSpill> whe
             }
 
             // process join
-            Iterator<Page> input = probePages.build().iterator();
             List<Page> actualPages = new ArrayList<>();
-            actualPages.addAll(toPagesPartial(joinOperator, singletonIterator(input.next())));
+
+            DriverContext ljoDriverContext = buildDrivers.get(0).getDriverContext().getPipelineContext().addDriverContext();
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: You'd rather add `PipelineContext` to `BuildSideSetup`

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 8]
```diff
@@ -283,18 +337,38 @@ public void testInnerJoinWithSpill(boolean probeHashEnabled, List<WhenSpill> whe
...
+                                          valuesOperatorFactory.createOperator(ljoDriverContext),
+                                          joinOperator,
+                                          pageBufferOperatorFactory.createOperator(ljoDriverContext));
 
             for (int i = 0; i < buildOperatorCount; i++) {
                 if (whenSpill.get(i) == WhenSpill.DURING_USAGE) {
-                    checkState(input.hasNext(), "spill requested DURING_USAGE, but input is kind of exhausted");
-                    triggerMemoryRevokingAndWait(buildSideSetup.buildOperators.get(i));
+                    buildSideSetup.buildOperators.get(i).getOperatorContext().requestMemoryRevoking();
+//                    triggerMemoryRevokingAndWait(buildSideSetup.buildOperators.get(i), taskStateMachine);
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: vestigial

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 9]
```diff
@@ -283,18 +337,38 @@ public void testInnerJoinWithSpill(boolean probeHashEnabled, List<WhenSpill> whe
...
                 }
             }
+            runDriver(executor, ljoDriver);
 
-            actualPages.addAll(toPages(joinOperator, input));
+            while (!ljoDriver.isFinished()) {
+                Thread.sleep(100);
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: Why not run LJO in foreground?
```
while (! ljoDriver.isFinished() /* or even better ! ljoDriver.isFinished() */) {
   ljoDriver.process();
}
```
?

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 10]
```diff
@@ -320,31 +394,57 @@ public void testInnerJoinWithSpill(boolean probeHashEnabled, List<WhenSpill> whe
         }
     }
 
+    private static void checkErrors(final TaskStateMachine taskStateMachine)
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: no need for `final`

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 11]
```diff
@@ -320,31 +394,57 @@ public void testInnerJoinWithSpill(boolean probeHashEnabled, List<WhenSpill> whe
...
+            throws Throwable
+    {
+        if (taskStateMachine.getFailureCauses().size() > 0) {
+            throw taskStateMachine.getFailureCauses().peek();
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java:401]: For the spill failure case, it's perfect. But if this triggers in non-failure-expecting test, then the fact you re-throw exception from another thread may be misleading. The stacktrace produced won't match the actual stack being traversed by the exception making source of exception un-traceable. 

Also, `taskStateMachine.getFailureCauses()` seems concurrent collection, so it's conceivable that `.peek()` returns `null`, thus consider

```
throw new RuntimeException(requireNotNull(taskStateMachine.getFailureCauses().peek()));
```

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 12]
```diff
@@ -320,31 +394,57 @@ public void testInnerJoinWithSpill(boolean probeHashEnabled, List<WhenSpill> whe
...
+    private enum WhenSpillFails
+    {
+        NEVER, SPILL_HBO, SPILL_LJO, UNSPILL_HBO, UNSPILL_LJO
+    }
+
     @DataProvider
     public Object[][] testInnerJoinWithSpillDataProvider()
     {
-        List<Object[]> result = new ArrayList<>();
+        return joinWithSpillParameters(true).stream().map(List::toArray).toArray(Object[][]::new);
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: Usually, one would put `.map` and `.toArray` on separate lines (retaining `.stream()` where it is)

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 13]
```diff
@@ -320,31 +394,57 @@ public void testInnerJoinWithSpill(boolean probeHashEnabled, List<WhenSpill> whe
...
+    }
+
+    private static List<List<Object>> joinWithSpillParameters(boolean allowNoSpill)
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: Is there any particular benefit from separating this out? Can't `product()` work directly on `Object[][]` instead?

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 14]
```diff
@@ -353,19 +453,20 @@ private static void revokeMemory(Operator operator)
         operator.finishMemoryRevoke();
     }
 
-    private static void triggerMemoryRevokingAndWait(Operator operator)
-            throws InterruptedException
+    private static <T> List<T> concat(List<T> initialElements, List<T> moreElements)
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: nit: put it below `product`

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 15]
```diff
@@ -971,7 +1073,13 @@ private static void runDriver(ExecutorService executor, Driver driver)
     {
         executor.execute(() -> {
             if (!driver.isFinished()) {
-                driver.process();
+                try {
+                    driver.process();
+                }
+                catch (PrestoException e) {
+                    driver.getDriverContext().failed(e);
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java:1080]: Why is this necessary? Doesn't driver call it by itself?

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 16]
```diff
@@ -1024,6 +1132,19 @@ public boolean filter(int leftPosition, Block[] leftBlocks, int rightPosition, B
     private static class DummySpillerFactory
             implements SingleStreamSpillerFactory
     {
+        private volatile boolean failSpill = false;
+        private volatile boolean failUnspill = false;
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: `false` is default

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 17]
```diff
@@ -1036,13 +1157,19 @@ public SingleStreamSpiller create(List<Type> types, SpillContext spillContext, L
...
+                    }
                     Iterators.addAll(spills, pageIterator);
                     return immediateFuture(null);
                 }
 
                 @Override
                 public Iterator<Page> getSpilledPages()
                 {
+                    if (failUnspill) {
+                        throw new PrestoException(GENERIC_INTERNAL_ERROR, "Unspill failed");
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: You could also test a case when iterator is returned successfully, but fails during e.g. `hasNext()`

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 18]
```diff
@@ -1058,6 +1185,9 @@ public long getSpilledPagesInMemorySize()
                 @Override
                 public ListenableFuture<List<Page>> getAllSpilledPages()
                 {
+                    if (failUnspill) {
+                        return immediateFailedFuture(new PrestoException(GENERIC_INTERNAL_ERROR, "Unspill failed"));
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: You could also test a case when `getAllSpilledPages()` throws directly.

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 19]
```diff
@@ -1036,13 +1157,19 @@ public SingleStreamSpiller create(List<Type> types, SpillContext spillContext, L
                 public ListenableFuture<?> spill(Iterator<Page> pageIterator)
                 {
                     checkState(writing, "writing already finished");
+                    if (failSpill) {
+                        return immediateFailedFuture(new PrestoException(GENERIC_INTERNAL_ERROR, "Spill failed"));
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: You could also test a case when `spill()` throws directly.

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 20]
```diff
@@ -283,18 +337,38 @@ public void testInnerJoinWithSpill(boolean probeHashEnabled, List<WhenSpill> whe
...
+            Driver ljoDriver = new Driver(ljoDriverContext,
+                                          valuesOperatorFactory.createOperator(ljoDriverContext),
+                                          joinOperator,
+                                          pageBufferOperatorFactory.createOperator(ljoDriverContext));
 
             for (int i = 0; i < buildOperatorCount; i++) {
                 if (whenSpill.get(i) == WhenSpill.DURING_USAGE) {
-                    checkState(input.hasNext(), "spill requested DURING_USAGE, but input is kind of exhausted");
-                    triggerMemoryRevokingAndWait(buildSideSetup.buildOperators.get(i));
+                    buildSideSetup.buildOperators.get(i).getOperatorContext().requestMemoryRevoking();
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: where is wait?

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 21]
```diff
@@ -914,7 +1016,7 @@ private BuildSideSetup setupBuildSide(
                 partitionCount,
                 new PagesIndex.TestingFactory(),
                 spillEnabled,
-                SINGLE_STREAM_SPILLER_FACTORY);
+                singleStreamSpillerFactory);
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: It looks like you started with separating preparatory changes but then you gave up and up everything in the last commit. For example, you could have this change, or changing LJO processing logic from "call it directly" to "wrap it with a driver" as a yet another preparatory commit.

File: presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java (hunk 3 above)
Comment (szymonm) [reply presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java:102]: I don't quite get it.
`ImmutableList.Builder()` takes all required parameters immediately in the ctor.
Here (before and after the change), Builder ctor takes required parameters while optional are provided later.

File: presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java (hunk 5 above)
Comment (szymonm) [reply presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java:49]: It looks like a convention in the file that `builder()` is not called directly, but it's wrapped in `createTaskContext`. Maybe we should also delete `createTaskContext(Executor notificationExecutor, ScheduledExecutorService yieldExecutor, Session session)` ?

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java (hunk 13 above)
Comment (szymonm) [reply presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: Tried that but it was looks more cleaner this way.

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java (hunk 20 above)
Comment (szymonm) [reply presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: I'm not sure still need it. Let's have a chat about that.

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java (hunk 21 above)
Comment (szymonm) [reply presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: True, will work on that.

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java (hunk 15 above)
Comment (szymonm) [reply presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java:1080]: Nope, it only marks that it needs distruction on the context and throws the expection up to be handled by TaskManager that is not mocked in the test.

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java (hunk 19 above)
Comment (szymonm) [reply presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: I'm wondering if those cases add much value. They clearly increase the number of combinations we are testing by on order of magnitude...

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java (hunk 6 above)
Comment (szymonm) [reply presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: I can't, because of using throw TaskStateMachine.peek() that has throwables.
